test/system/src/test/java/io/pravega/test/system/MultiReaderWriterWithFailOverTest.java

```diff
@@ -0,0 +1,427 @@
+                        new JavaSerializer<Long>(),
+                        EventWriterConfig.builder().build());
+                writerList.add(writer);
+            }
+
+            //create a reader group
+            log.info("Creating Reader group : {}", readerGroupName);
+            log.info("Scope passed to readergroup manager {}", scope);
+
+            readerGroupManager.createReaderGroup(readerGroupName, ReaderGroupConfig.builder().startingTime(0).build(),
+                    Collections.singleton(STREAM_NAME));
+            log.info("Reader group name {} ", readerGroupManager.getReaderGroup(readerGroupName).getGroupName());
+            log.info("Reader group scope {}", readerGroupManager.getReaderGroup(readerGroupName).getScope());
+            log.info("Online readers {}", readerGroupManager.getReaderGroup(readerGroupName).getOnlineReaders());
+
+            //create readers
+            log.info("Creating {} readers", NUM_READERS);
+            List<EventStreamReader<Long>> readerList = new ArrayList<>();
+            String readerName = "reader" + new Random().nextInt(Integer.MAX_VALUE);
+            log.info("Scope that is seen by readers {}", scope);
+            //start reading events
+            for (int i = 0; i < NUM_READERS; i++) {
+                log.info("Starting reader{}", i);
+                log.info("Creating reader with id {}", readerName + i);
+                final EventStreamReader<Long> reader = clientFactory.createReader(readerName + i,
+                        readerGroupName,
+                        new JavaSerializer<Long>(),
+                        ReaderConfig.builder().build());
+                readerList.add(reader);
+            }
+
+            // start writing asynchronously
+            List<CompletableFuture<Void>> writerFutureList = new ArrayList<>();
+            writerList.forEach(writer -> {
+                CompletableFuture<Void> writerFuture = startWriting(eventData, writer, stopWriteFlag);
+                writerFutureList.add(writerFuture);
+            });
+
+            //start reading asynchronously
+            List<CompletableFuture<Void>> readerFutureList = new ArrayList<>();
+            readerList.forEach(reader -> {
+                CompletableFuture<Void> readerFuture = startReading(eventsReadFromPravega, eventData, eventReadCount, stopReadFlag, reader);
+                writerFutureList.add(readerFuture);
+            });
+
+            //perform the scaling operations
+            performScaling();
+
+            //set the stop write flag to true
+            log.info(" stop write flag {}", stopWriteFlag);
+            stopWriteFlag.set(true);
+
+            //wait for writers completion
+            log.info("wait for writers exceution to complete");
+            FutureHelpers.allOf(writerFutureList);
+
+            //set the stop read flag to true
+            stopReadFlag.set(true);
+
+            // wait for reads = writes
+            synchronized (waitCondition) {
```

Comment:
Since we are using the stop flag and doing FutureHelpers.allOf(readerFutureList) for synchronization, we don't need wait/notify mechanism. Review if this is still required and remove it otherwise.